### PR TITLE
[OC-1584] Usercard : permit to have a restricted list of recipient

### PR DESCRIPTION
**Pull request by quinarygio**
Release notes

In Tasks section:
[OC-1584] Usercard : permit to have a restricted list of recipients  depending on the state and process
Note: in bundle config.json added in state definition  a "recipientList" field to configure a list of entities and optionally the levels of connection of related childs to be included in recipient list. Level 0 meaning the entity itself, level 1 for first level childs, level 2 for 2nd level connections, etc.

[OC-1584]: https://opfab.atlassian.net/browse/OC-1584

### CODE REVIEW

File: ui/main/src/app/model/processes.model.ts [hunk 1]
```diff
@@ -69,9 +69,19 @@ export class UserCard {
...
         readonly endDateVisible?: boolean,
         readonly lttdVisible?: boolean,
+        readonly recipientList?: Recipient[]
     ) {
     }
 }
+
+export class Recipient {
+    constructor(
+        readonly id?: string,
```
Comment (AlexGuironnetRTE) [inline ui/main/src/app/model/processes.model.ts:79]: Is the id really meant to be optional ?

File: services/core/businessconfig/src/main/java/org/opfab/businessconfig/model/UserCardData.java [hunk 2]
```diff
@@ -11,6 +11,8 @@
 
 package org.opfab.businessconfig.model;
 
+import java.util.List;
+import javax.validation.Valid;
```
Comment (AlexGuironnetRTE) [inline services/core/businessconfig/src/main/java/org/opfab/businessconfig/model/UserCardData.java:15]:
```suggestion
```
Sonar says this import is not used.

**Review by AlexGuironnetRTE — CHANGES_REQUESTED** [review]
Hello,

Documentation is clear, I have a few remarks on the code (see comments), the given example works as intended but the following edge case doesn't:
This config: 
`
"recipientList" : [{"id": "ALLCONTROLROOMS", "levels": [1,0]}, {"id": "ENTITY4"}, {"id": "ENTITY2"}, {"id": "entityKarate1", "levels":  [1]}]`

yields

image

Control Room 2 is duplicated.

File: services/core/businessconfig/src/main/modeling/swagger.yaml [hunk 3]
```diff
@@ -535,4 +535,20 @@ definitions:
...
\ No newline at end of file
+        type: boolean
+      recipientList:
+        type: array
+        items:
+          $ref: '#/definitions/Recipient'
+  Recipient:
+    description: >- 
+      Object containing the id of the recipient entity and an optional list of connection levels 
+      with 0 meaning the entity itself, 1 for first level childs, 2 for 2nd level connections, etc.
```
[the remark this answers is not in the document]
Comment (quinarygio) [reply services/core/businessconfig/src/main/modeling/swagger.yaml]: Done

> Hello,
> 
> Documentation is clear, I have a few remarks on the code (see comments), the given example works as intended but the following edge case doesn't:
> This config:
> ` "recipientList" : [{"id": "ALLCONTROLROOMS", "levels": [1,0]}, {"id": "ENTITY4"}, {"id": "ENTITY2"}, {"id": "entityKarate1", "levels": [1]}]`
> 
> yields
> 
> image
> 
> Control Room 2 is duplicated.

Fixed.

File: src/docs/asciidoc/reference_doc/user_cards.adoc [hunk 4]
```diff
@@ -30,7 +30,8 @@ For example :
       "severityVisible" : true,
       "startDateVisible" : true,
       "endDateVisible" : true,
-      "lttdVisible" : false
+      "lttdVisible" : false,
+      "recipientList" : [{"id": "ENTITY1", "levels": [0,1]}, {"id": "ENTITY4"}]
```
[the remark this answers is not in the document]
Comment (quinarygio) [reply src/docs/asciidoc/reference_doc/user_cards.adoc]: Done

File: src/docs/asciidoc/reference_doc/user_cards.adoc [hunk 5]
```diff
@@ -135,6 +136,32 @@ or `Write` for the chosen process and state. For example:
...
+    },
+    "templateName": "message",
+    "styles": [],
+    "acknowledgmentAllowed": "Always"
+  }
+}
+....
+
+In this example the list of available recipients will contain: "ENTITY1" (level 0), all the first level childs of "ENTITY1" (level 1) and "ENTITY4".
```
[the remark this answers is not in the document]
Comment (quinarygio) [reply src/docs/asciidoc/reference_doc/user_cards.adoc]: Done

File: src/docs/asciidoc/reference_doc/user_cards.adoc [hunk 6]
```diff
@@ -135,6 +136,32 @@ or `Write` for the chosen process and state. For example:
...
+"states": {
+  "messageState": {
+    "name": "message.title",
+    "userCard" : {
+      "template" : "usercard_message",
+      "severityVisible" : true,
+      "startDateVisible" : true,
+      "endDateVisible" : true,
+      "lttdVisible" : false,
+      "recipientList" : [{"id": "ENTITY1", "levels": [0,1]}, {"id": "ENTITY4"}]
```
[the remark this answers is not in the document]
Comment (quinarygio) [reply src/docs/asciidoc/reference_doc/user_cards.adoc]: Done

File: ui/main/src/app/model/processes.model.ts (hunk 1 above)
Comment (quinarygio) [reply ui/main/src/app/model/processes.model.ts]: You are right,.. Fixed

File: services/core/businessconfig/src/main/java/org/opfab/businessconfig/model/UserCardData.java (hunk 2 above)
Comment (quinarygio) [reply services/core/businessconfig/src/main/java/org/opfab/businessconfig/model/UserCardData.java]: Done

**Review by AlexGuironnetRTE — APPROVED** [review]
Changes and tests OK

**Pull request merged**